File: 06-observability/05-red-and-use/python/pool_saturation.py
"""
Layer 6 Topic 5 - Utilization is not saturation: the connection pool, in Python.

Why Python: SQLAlchemy's pool is the bottom rung of the ladder. It exposes
`checkout` and `checkin` events and a `status()` string, and there is no timer
anywhere in it. The number you need during an incident -- how long a request
waited for a connection -- does not exist until you write it. Worse, and this
is Topic 2's finding arriving as a metric: the wait happens BEFORE any
statement is issued, so it produces no span either. Auto-instrumentation gives
you the server span and the DB spans and nothing in between.

What this program does
----------------------
A real pool -- five permits handed out FIFO, which is what `pool_size=5,
max_overflow=0` is -- under a real ramp of real threads, each holding a
connection for a fixed service time and doing a little application work in
between. Every number below is measured in this process; nothing is modelled.

Three views of the same ramp, printed side by side:

  1. UTILIZATION  -- checked-out / size. What every default dashboard shows.
     It pins at 100% and stops carrying information, which is exactly when the
     incident starts.
  2. SATURATION   -- queue depth and checkout WAIT TIME. Unbounded, still
     moving, and the thing that correlates with what users feel.
  3. What SQLAlchemy gives you for free: `pool.status()`, a string, sampled.
     Compare it against the ground truth in the same row.

Then two closing sections:

  * the hand-written instrumentation that produces the metric that does not
    exist by default: a `checkout` event handler, a timestamp, a subtraction,
    a histogram, and `db.client.connection.count{state=used|idle}` beside it.
  * one slow request's span timeline, showing the pool wait as a GAP with no
    span in it -- the reason this is invisible in a trace UI until you go and
    instrument checkout yourself.

What to look for in the output
------------------------------
The step where utilization hits 100% and the step where checkout waits begin.
They are not the same step, and the order they arrive in is the finding: by the
time saturation is visible in the utilization column, the column has been
saturated for a while and stopped changing. Then read the last column of the
ramp table: the mean is not the p99, and the pool is exactly the place where
the mean is the least interesting statistic.
"""
import sys
import threading
import time
from collections import deque

POOL_SIZE = 5          # SQLAlchemy: pool_size=5, max_overflow=0
SERVICE_TIME = 0.005   # 5ms holding the connection: a fast indexed query
THINK_TIME = 0.010     # 10ms of application work between queries
STEP_SECONDS = 1.0
POLL_INTERVAL = 0.25   # the dashboard's scrape, scaled to the step length
WORKER_STEPS = [2, 5, 10, 25, 60, 120]   # the lab's ramp, compressed


class InstrumentedPool:
    """A connection pool with the instrumentation SQLAlchemy does not ship.

    The `checkout`/`checkin` hooks are exactly SQLAlchemy's events; everything
    inside them is the code you have to write yourself. That is the lesson: the
    hooks exist, the metric does not.
    """

    def __init__(self, size):
        self.size = size
        # A FIFO queue of waiters, not a Semaphore: SQLAlchemy's QueuePool
        # hands the connection to the longest waiter, and an unfair pool would
        # produce a starvation distribution rather than a queueing one.
        self._available = size
        self._waiters = deque()
        self._lock = threading.Lock()
        self.checked_out = 0
        self.waiting = 0
        # Instrumentation -- none of this exists by default.
        self.wait_times = []          # the histogram nobody ships
        self.max_waiting = 0
        self.checkouts = 0
        self.timeouts = 0

    def acquire(self, timeout=30.0):
        """Returns the wait in seconds, or None on pool_timeout."""
        start = time.perf_counter()
        with self._lock:
            if self._available and not self._waiters:
                self._available -= 1
                self.checked_out += 1
                self.checkouts += 1
                self.wait_times.append(0.0)
                return 0.0
            ticket = threading.Event()
            self._waiters.append(ticket)
            self.waiting += 1
            self.max_waiting = max(self.max_waiting, self.waiting)

        granted = ticket.wait(timeout)
        wait = time.perf_counter() - start

        with self._lock:
            self.waiting -= 1
            if not granted:
                # pool_timeout: SQLAlchemy raises TimeoutError here, and this
                # is the one pool event that DOES show up in your logs.
                self._waiters.remove(ticket)
                self.timeouts += 1
                return None
            self.checked_out += 1
            self.checkouts += 1
            self.wait_times.append(wait)   # <- the metric that does not exist
        return wait

    def release(self):
        with self._lock:
            self.checked_out -= 1
            if self._waiters:
                self._waiters.popleft().set()   # longest waiter first
            else:
                self._available += 1

    # --- the two things SQLAlchemy actually gives you -----------------------

    def status(self):
        """SQLAlchemy's QueuePool.status(). A string. Sampled, not recorded."""
        return ("Pool size: %d  Connections in pool: %d  Current Overflow: %d  "
                "Current Checked out connections: %d"
                % (self.size, self.size - self.checked_out, 0, self.checked_out))

    def utilization(self):
        return self.checked_out / self.size


def percentile(values, q):
    """Nearest-rank percentile. No numpy on this machine, and none needed."""
    if not values:
        return 0.0
    ordered = sorted(values)
    index = max(0, min(len(ordered) - 1, int(q * len(ordered) + 0.999999) - 1))
    return ordered[index]


def run_step(pool, workers, seconds):
    """`workers` threads hammering the pool for `seconds`. Returns measurements."""
    stop = time.perf_counter() + seconds
    first_index = len(pool.wait_times)
    pool.max_waiting = 0
    samples = deque()          # the 15-second-poll dashboard, sped up
    latencies = []
    lat_lock = threading.Lock()

    def worker():
        local = []
        while time.perf_counter() < stop:
            request_start = time.perf_counter()
            wait = pool.acquire()
            if wait is None:
                continue
            time.sleep(SERVICE_TIME)   # holding the connection: the query
            pool.release()
            local.append(time.perf_counter() - request_start)
            time.sleep(THINK_TIME)     # application work, connection returned
        with lat_lock:
            latencies.extend(local)

    def sampler():
        # A dashboard scraping the gauge. Every sample is a moment; everything
        # between two samples is invisible to it.
        while time.perf_counter() < stop:
            samples.append((pool.utilization(), pool.waiting))
            time.sleep(POLL_INTERVAL)

    threads = [threading.Thread(target=worker, daemon=True) for _ in range(workers)]
    sampler_thread = threading.Thread(target=sampler, daemon=True)
    sampler_thread.start()
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    sampler_thread.join()

    waits = pool.wait_times[first_index:]
    sampled_util = max((s[0] for s in samples), default=0.0)
    sampled_queue = max((s[1] for s in samples), default=0)
    return {
        "workers": workers,
        "requests": len(waits),
        "sampled_util": sampled_util,
        "true_max_queue": pool.max_waiting,
        "sampled_queue": sampled_queue,
        "wait_mean": sum(waits) / len(waits) if waits else 0.0,
        "wait_p50": percentile(waits, 0.50),
        "wait_p99": percentile(waits, 0.99),
        "lat_p99": percentile(latencies, 0.99),
        "timeouts": pool.timeouts,
    }


def main():
    print("Layer 6 Topic 5 - utilization vs saturation, on a Python connection pool")
    print("python %s   pool_size=%d, max_overflow=0, service time %.0f ms"
          % (sys.version.split()[0], POOL_SIZE, SERVICE_TIME * 1000))
    print("=" * 78)
    print()

    pool = InstrumentedPool(POOL_SIZE)
    rows = []
    for workers in WORKER_STEPS:
        rows.append(run_step(pool, workers, STEP_SECONDS))

    print("--- The ramp: one pool, six concurrency levels, everything measured ---")
    print()
    print("           |  USE: utilization  |        USE: saturation        |   RED   ")
    print("  workers  |  polled  in use    |  max queued   wait p50 / p99  |  req p99")
    print("  ---------+--------------------+-------------------------------+---------")
    for r in rows:
        print("  %7d  |  %5.0f%%   %2d of %d   |  %10d   %5.1f / %6.1f ms |  %6.1f ms"
              % (r["workers"], 100 * r["sampled_util"],
                 round(r["sampled_util"] * POOL_SIZE), POOL_SIZE,
                 r["true_max_queue"], 1000 * r["wait_p50"], 1000 * r["wait_p99"],
                 1000 * r["lat_p99"]))
    print()

    first_pinned = next((r for r in rows if r["sampled_util"] >= 0.999), None)
    first_waited = next((r for r in rows if r["wait_p99"] > 0.001), None)
    pinned_at = first_pinned["workers"] if first_pinned else None
    waited_at = first_waited["workers"] if first_waited else None
    print("  polled utilization first reads 100%%   at %s workers"
          % (pinned_at if pinned_at else "never"))
    print("  checkout wait p99 first exceeds 1ms   at %s workers"
          % (waited_at if waited_at else "never"))
    print()
    if pinned_at and waited_at and waited_at < pinned_at:
        print("  On this run the queue formed BEFORE the utilization gauge ever")
        print("  read 100%. That is not a paradox and it is not luck: utilization")
        print("  is sampled, and a pool that is full for 40 ms at a time is full")
        print("  in between two scrapes. The queue, meanwhile, was recorded on")
        print("  every checkout, so it had nothing to miss.")
    elif pinned_at and waited_at and waited_at == pinned_at:
        print("  Both crossed at the same step, which is the least interesting")
        print("  of the three orders and still makes the point: after this step")
        print("  the utilization column never changes again, and every later row")
        print("  is described entirely by the saturation columns.")
    elif pinned_at and waited_at:
        print("  Utilization pinned first, then stopped moving. Everything after")
        print("  that point is described only by the saturation columns.")
    print()
    print("  Either way, read the util% column downward: it reaches its maximum")
    print("  and stays there, with no room left to describe anything, while the")
    print("  wait columns keep climbing with no upper bound. That is the whole")
    print("  distinction, and it is why the green number is the one that lies.")
    print()

    print("--- What SQLAlchemy gives you for free ---")
    print()
    print("  pool.status():")
    print("    %s" % pool.status())
    print()
    print("  That is the entire built-in observable: a string, describing this")
    print("  instant, with no history and no timing. Nothing in SQLAlchemy counts")
    print("  how long anybody waited. Ask it 'what was the p99 checkout wait during")
    print("  the incident' and there is no object in the library that holds an")
    print("  answer.")
    print()

    print("--- What a 15-second-poll dashboard would have seen ---")
    print()
    print("  %-12s %-16s %-16s" % ("workers", "true max queue", "polled max queue"))
    for r in rows:
        missed = "" if r["sampled_queue"] >= r["true_max_queue"] else "   <- missed"
        print("  %-12d %-16d %-16d%s"
              % (r["workers"], r["true_max_queue"], r["sampled_queue"], missed))
    print()
    print("  This program polls every %.0f ms and still misses peaks. A real"
          % (1000 * POLL_INTERVAL))
    print("  dashboard polls every 15 seconds. A 400ms queue between two scrapes")
    print("  did not happen, as far as your graph is concerned. A gauge is a")
    print("  sample; a histogram is a record. Saturation needs the record.")
    print()

    print("--- The instrumentation that produces the missing metric ---")
    print()
    print("  Fifteen lines, in the events SQLAlchemy already gives you:")
    print()
    print("    @event.listens_for(engine, 'checkout')")
    print("    def on_checkout(dbapi_conn, conn_record, conn_proxy):")
    print("        wait = time.perf_counter() - conn_record.info['requested_at']")
    print("        checkout_wait_histogram.record(wait)          # the missing one")
    print("        pool_state.add(1, {'state': 'used'})")
    print()
    print("  and the semconv names to publish them under:")
    print("    db.client.connection.count{state=\"used\"|\"idle\"}   (gauge)")
    print("    db.client.connection.wait_time                     (histogram)")
    print()
    all_waits = pool.wait_times
    print("  What that histogram holds after this run, from %s checkouts:"
          % f"{len(all_waits):,}")
    print("    p50   %7.2f ms" % (1000 * percentile(all_waits, 0.50)))
    print("    p95   %7.2f ms" % (1000 * percentile(all_waits, 0.95)))
    print("    p99   %7.2f ms" % (1000 * percentile(all_waits, 0.99)))
    print("    max   %7.2f ms" % (1000 * max(all_waits)))
    print("    mean  %7.2f ms   <- the only one a counter-pair could have given you"
          % (1000 * sum(all_waits) / len(all_waits)))
    print()
    print("  The mean is %.1fx smaller than the p99. Pool waits are exactly the"
          % ((percentile(all_waits, 0.99) / (sum(all_waits) / len(all_waits)))
             if all_waits else 0))
    print("  distribution where the mean is the least interesting statistic, which")
    print("  is worth remembering when you get to the Go program in this topic and")
    print("  find that a mean is all `database/sql` can give you.")
    print()

    print("--- Why you cannot see this in a trace (Topic 2's finding, as a metric) ---")
    print()
    slow = max(rows, key=lambda r: r["wait_p99"])
    wait_ms = 1000 * slow["wait_p99"]
    query_ms = SERVICE_TIME * 1000
    total = wait_ms + query_ms
    scale = 60.0 / total if total else 1.0
    print("  One request at the p99 of the %d-worker step, drawn to scale:" % slow["workers"])
    print()
    print("    server span   |%s|  %.1f ms" % ("=" * int(total * scale), total))
    print("    (no span)     |%s%s|  %.1f ms  <- pool wait"
          % (" " * 0, "?" * max(1, int(wait_ms * scale)), wait_ms))
    print("    db span       |%s%s|  %.1f ms"
          % (" " * max(1, int(wait_ms * scale)), "=" * max(1, int(query_ms * scale)),
             query_ms))
    print()
    print("  %.0f%% of that request is a gap with no span in it. SQLAlchemy checks a"
          % (100 * wait_ms / total))
    print("  connection out BEFORE issuing any statement, so there is nothing for")
    print("  auto-instrumentation to hook: no statement, no span. In Tempo this")
    print("  reads as dead time between the server span starting and the first DB")
    print("  span, and the natural conclusion -- 'Python was slow' -- is wrong.")
    print()
    print("  Instrument checkout, and the gap gets a name.")


if __name__ == "__main__":
    main()

File: 06-observability/05-red-and-use/nodejs/pool_saturation.js
/*
 * Layer 6 Topic 5 - Utilization is not saturation: the connection pool, in Node.
 *
 * Why Node.js: node-postgres is the second rung of the ladder. `pool.totalCount`,
 * `pool.idleCount` and `pool.waitingCount` are real, documented, free -- and all
 * three are INSTANTANEOUS. `waitingCount` is the queue depth right now, not a
 * distribution and not a total. Poll it every 15 seconds and a 400ms queue
 * between two polls did not happen, as far as your dashboard is concerned.
 *
 * That is strictly better than Python's nothing, and strictly worse than it
 * looks. Knowing which is the point of this program, and the gap is measured
 * here rather than asserted: every checkout is timed (ground truth) while a
 * sampler polls the gauge on a fixed interval, and the two are printed side by
 * side.
 *
 * The pool below is node-postgres's shape: a fixed `max`, a FIFO queue of
 * pending acquisitions, and the same three counters exposed under the same
 * names. No pg module is installed on this machine and none is needed -- the
 * observable being studied is the stats surface, not the wire protocol.
 *
 * What to look for in the output
 * ------------------------------
 * 1. The ramp table. Utilization pins at 100% and stops carrying information;
 *    queue depth and wait time keep climbing with no upper bound.
 * 2. The "polled vs true" table. Same run, same pool: what the gauge saw and
 *    what actually happened. The gap is the cost of sampling a queue.
 * 3. The last section: the single-thread consequence. Every one of those
 *    waiting requests is a live promise on one event loop, so pool saturation
 *    in Node is not just slow requests -- it is memory, and it is the same
 *    event loop that has to run your health check.
 */

'use strict';

const POOL_MAX = 5;          // node-postgres: new Pool({ max: 5 })
const SERVICE_TIME_MS = 5;   // holding the connection: a fast indexed query
const THINK_TIME_MS = 10;    // application work between queries
const STEP_MS = 1000;
const POLL_INTERVAL_MS = 250; // the dashboard's scrape, scaled to the step
const CONCURRENCY_STEPS = [2, 5, 10, 25, 60, 120];

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

// ---------------------------------------------------------------------------
// node-postgres's Pool, cut down to its stats surface plus the instrumentation
// it does not ship (every `waitMs` below is ours, not theirs).
// ---------------------------------------------------------------------------

class Pool {
  constructor(max) {
    this.max = max;
    this._inUse = 0;
    this._queue = []; // FIFO, exactly like pg's pending-acquire queue
    // --- what node-postgres exposes -------------------------------------
    // totalCount, idleCount, waitingCount: all instantaneous, all counts.
    // --- what we added --------------------------------------------------
    this.waitSamples = [];
    this.maxWaiting = 0;
    this.checkouts = 0;
  }

  get totalCount() { return this.max; }
  get idleCount() { return this.max - this._inUse; }
  get waitingCount() { return this._queue.length; }
  get utilization() { return this._inUse / this.max; }

  connect() {
    const started = process.hrtime.bigint();
    if (this._inUse < this.max && this._queue.length === 0) {
      this._inUse += 1;
      this.checkouts += 1;
      this.waitSamples.push(0);
      return Promise.resolve(0);
    }
    return new Promise((resolve) => {
      this._queue.push(() => {
        const waitMs = Number(process.hrtime.bigint() - started) / 1e6;
        this._inUse += 1;
        this.checkouts += 1;
        this.waitSamples.push(waitMs); // <- the record node-postgres has no room for
        resolve(waitMs);
      });
      if (this._queue.length > this.maxWaiting) this.maxWaiting = this._queue.length;
    });
  }

  release() {
    this._inUse -= 1;
    const next = this._queue.shift();
    if (next) next();
  }
}

function percentile(values, q) {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const index = Math.min(sorted.length - 1, Math.max(0, Math.ceil(q * sorted.length) - 1));
  return sorted[index];
}

// ---------------------------------------------------------------------------
// One step of the ramp: `concurrency` in-flight request loops for STEP_MS.
// ---------------------------------------------------------------------------

async function runStep(pool, concurrency) {
  const deadline = Date.now() + STEP_MS;
  const firstSample = pool.waitSamples.length;
  pool.maxWaiting = 0;

  const latencies = [];
  const polled = { util: 0, queue: 0, samples: 0 };
  let lagMax = 0;

  const poller = setInterval(() => {
    // This is the scrape. It reads the same three properties a Prometheus
    // exporter for node-postgres reads, and it can only see this instant.
    polled.util = Math.max(polled.util, pool.utilization);
    polled.queue = Math.max(polled.queue, pool.waitingCount);
    polled.samples += 1;
  }, POLL_INTERVAL_MS);

  // Event-loop lag: schedule a 0ms timer and see how late it actually runs.
  // On a saturated pool this is the number that decides whether your health
  // check answers.
  const lagTimer = setInterval(() => {
    const expected = Date.now() + 20;
    setTimeout(() => {
      lagMax = Math.max(lagMax, Date.now() - expected);
    }, 20);
  }, 50);

  async function requestLoop() {
    while (Date.now() < deadline) {
      const started = process.hrtime.bigint();
      await pool.connect();
      await sleep(SERVICE_TIME_MS); // holding the connection
      pool.release();
      latencies.push(Number(process.hrtime.bigint() - started) / 1e6);
      await sleep(THINK_TIME_MS);   // connection returned, app work continues
    }
  }

  await Promise.all(Array.from({ length: concurrency }, requestLoop));
  clearInterval(poller);
  clearInterval(lagTimer);

  const waits = pool.waitSamples.slice(firstSample);
  return {
    concurrency,
    requests: waits.length,
    polledUtil: polled.util,
    polledQueue: polled.queue,
    polls: polled.samples,
    trueMaxQueue: pool.maxWaiting,
    waitP50: percentile(waits, 0.5),
    waitP99: percentile(waits, 0.99),
    waitMean: waits.reduce((a, b) => a + b, 0) / (waits.length || 1),
    latP99: percentile(latencies, 0.99),
    lagMax,
  };
}

async function main() {
  console.log('Layer 6 Topic 5 - utilization vs saturation, on a Node.js connection pool');
  console.log(`node ${process.version}   max=${POOL_MAX}, service time ${SERVICE_TIME_MS} ms, `
    + `scrape every ${POLL_INTERVAL_MS} ms`);
  console.log('='.repeat(78));
  console.log();

  const pool = new Pool(POOL_MAX);
  const rows = [];
  for (const concurrency of CONCURRENCY_STEPS) {
    rows.push(await runStep(pool, concurrency));
  }

  console.log('--- The ramp: one pool, six concurrency levels, everything measured ---');
  console.log();
  console.log('              |  USE: utilization  |      USE: saturation      |   RED   ');
  console.log('  in flight   |  polled  in use    |  max queued  wait p50/p99 |  req p99');
  console.log('  ------------+--------------------+---------------------------+---------');
  for (const r of rows) {
    console.log(
      `  ${String(r.concurrency).padStart(9)}   |  ${(100 * r.polledUtil).toFixed(0).padStart(4)}%`
      + `   ${`${Math.round(r.polledUtil * POOL_MAX)} of ${POOL_MAX}`.padEnd(9)}|`
      + `  ${String(r.trueMaxQueue).padStart(10)}  ${r.waitP50.toFixed(1).padStart(5)}/`
      + `${r.waitP99.toFixed(1).padStart(6)} ms |  ${r.latP99.toFixed(1).padStart(6)} ms`,
    );
  }
  console.log();

  const pinned = rows.find((r) => r.polledUtil >= 0.999);
  const waited = rows.find((r) => r.waitP99 > 1);
  console.log(`  polled utilization first reads 100%   at ${pinned ? pinned.concurrency : 'never'} in flight`);
  console.log(`  checkout wait p99 first exceeds 1ms   at ${waited ? waited.concurrency : 'never'} in flight`);
  console.log();
  console.log('  Once utilization reaches 100% it stops moving. Queue depth and wait');
  console.log('  time have no upper bound, so they are the only columns still saying');
  console.log('  anything about how bad it is.');
  console.log();

  console.log('--- What node-postgres gives you for free, and what it costs you ---');
  console.log();
  console.log(`  pool.totalCount    ${pool.totalCount}      (max, a constant)`);
  console.log(`  pool.idleCount     ${pool.idleCount}      (this instant)`);
  console.log(`  pool.waitingCount  ${pool.waitingCount}      (this instant)`);
  console.log();
  console.log('  Three counts, no timings, no history. To answer "what was the p99');
  console.log('  checkout wait during the incident" you need a record of every');
  console.log('  checkout, and the library keeps none -- so the answer has to come');
  console.log('  from code you wrote, exactly as in the Python program.');
  console.log();

  console.log('--- What the scrape saw, against what happened ---');
  console.log();
  console.log('  in flight   scrapes   true max queue   polled max queue   missed by');
  for (const r of rows) {
    const missed = r.trueMaxQueue - r.polledQueue;
    console.log(
      `  ${String(r.concurrency).padStart(9)}   ${String(r.polls).padStart(7)}`
      + `   ${String(r.trueMaxQueue).padStart(14)}   ${String(r.polledQueue).padStart(16)}`
      + `   ${missed > 0 ? `${missed} requests` : 'nothing'}`,
    );
  }
  console.log();
  console.log(`  This program scrapes every ${POLL_INTERVAL_MS} ms. Production scrapes every`);
  console.log('  15 seconds -- sixty times less often. A gauge answers "how bad is it');
  console.log('  right now"; it cannot answer "how bad did it get", and an incident is');
  console.log('  a question of the second kind.');
  console.log();

  const allWaits = pool.waitSamples;
  console.log('--- The histogram you have to add, and what it holds ---');
  console.log();
  console.log(`  from ${allWaits.length.toLocaleString()} checkouts across the whole ramp:`);
  console.log(`    p50   ${percentile(allWaits, 0.5).toFixed(2).padStart(8)} ms`);
  console.log(`    p95   ${percentile(allWaits, 0.95).toFixed(2).padStart(8)} ms`);
  console.log(`    p99   ${percentile(allWaits, 0.99).toFixed(2).padStart(8)} ms`);
  console.log(`    max   ${Math.max(...allWaits).toFixed(2).padStart(8)} ms`);
  console.log();
  console.log('  Publish it as db.client.connection.wait_time, next to');
  console.log('  db.client.connection.count{state="used"|"idle"} built from the two');
  console.log('  counts the library does give you. That pair is the USE column for a');
  console.log('  pool: one number for utilization, one for saturation.');
  console.log();

  console.log('--- The Node-specific part: one thread, and every waiter is on it ---');
  console.log();
  console.log('  in flight   max queued   event-loop lag (max)');
  for (const r of rows) {
    console.log(
      `  ${String(r.concurrency).padStart(9)}   ${String(r.trueMaxQueue).padStart(10)}`
      + `   ${`${r.lagMax} ms`.padStart(20)}`,
    );
  }
  console.log();
  console.log('  A queued request in Python is a blocked OS thread; in Node it is a');
  console.log('  pending promise on the one thread that also runs your health check,');
  console.log('  your metrics endpoint and every other request. That makes the honest');
  console.log('  measurement cheap to take (Topic 2) and makes saturation everybody\'s');
  console.log('  problem at once: when the pool backs up, the queue is not somewhere');
  console.log('  else, it is in the same loop as everything you would use to find out.');
}

main();
